docs/architecture/adr-014-versioned-namespaces.md

```diff
@@ -0,0 +1,192 @@
+  1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, // namespace ID
+  1,          // info byte (sequence start indicator = true)
+  0, 0, 0, 3, // sequence length
+  1, 2, 3, // blob data
+  // 0 padding until share is full
+ }
+```
+
+Pros
+
+- Supports variable length namespaces without a namespace version bump
+
+Cons
+
+- 2 bytes of metadata per share
+
+Notes
+
+- Same as Option A: “Additional changes are necessary…”
+- Option B is implementable at a later time if we adopt Option A.
+
+### Option C: `Share Version | Namespace ID`
+
+Inspired by [CIDs](https://github.com/multiformats/cid#design-considerations), consider encoding a share version **at the start of the share** to ensure the share format itself can evolve. In other words, the share version moves from after the namespace to before the namespace. The share version can then be used to version changes to the namespace and universal share prefix.
+
+```go
+optionC := []byte{
+  0,                      // share version
+  1, 2, 3, 4, 5, 6, 7, 8, // namespace ID
+  1,          // info byte (only contains sequence start indicator = true)
+  0, 0, 0, 3, // sequence length
+  1, 2, 3, // blob data
+  // 0 padding until share is full
+ }
+```
+
+Pros
+
+- Fewer fields need to be specified. In options A and B, an end-user needs to specify `shareVersion=0` and `namespaceVersion=0`. With this option, they only need to specify  `shareVersion=0`.
+- The existing info byte can be renamed to the sequence start indicator for clarity. Alternatively it can remain an info byte with 7 available bits for future metadata.
+
+Cons
+
+- Since the namespace and the universal share prefix are versioned via the same field, a breaking change to either mandates a share version bump.
+- Roll-ups can’t adopt a new share version without also moving the namespace that they read/write from.
+
+Notes
+
+- With this option, the share version is prefixed to the namespace prior to pushing to the NMT. When a user queries a namespace, they must also specify a share version.
+
+## Implementation Details
+
+Regardless of the option chosen, end-users must explicitly specify the namespace ID version (and/or) share version they intend on using in their PFB transaction.
+
+Since the NMT needs to be aware of the version byte, Option A is equivalent to increasing the namespace ID size to `9` bytes and then constraining the namespaces available for use to namespaces that have a leading `0` byte. In other words:
+
+- MinReservedNamespace: []byte{0, 0, 0, 0, 0, 0, 0, 0, 1}
+- MaxReservedNamespace: []byte{0, 0, 0, 0, 0, 0, 0, 0, 255}
+- MinBlobNamespace: []byte{0, 0, 0, 0, 0, 0, 0, 1, 0}
+- MaxBlobNamespace: []byte{0, 255, 255, 255, 255, 255, 255, 255, 255}
+
+When a user creates a PFB, concatenate the namespace version with the namespace ID to derive the namespace that is pushed to the NMT. Option C is similar to Option A, if we constrain the available share versions at mainnet to `0`.
+
+## Open Questions
```

Comment:
what version byte should we use for tail padding shares and parity shares? I assume we need to use the largest version?

Reply:
Good question. Given we adopt **Option A**.

### Option A1 
Use the maximum available namespace version. For example, at mainnet launch if we constrained the namespace versions to only `0`, then the tail padding share would be:

- namespace version: `0`
- namespace ID: `namespace.ID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE}`

At a later date when we support namespace versions `0` and `1`, the tail padding share would be:

- namespace version: `1`
- namespace ID: `namespace.ID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE}`

### Option A2
Use the maximum namespace version regardless of supported namespace versions. So a tail padding share would be:

- namespace version: `255`
- namespace ID: `namespace.ID{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFE}`

Reply:
A2 makes the most sense to me, since we won't have to continuously update it

Reply:
its kinda weird since the only reason we have to do this is because nmt doesn't know about the version byte

Reply:
In this case, I expect we'll prefix the namespace ID with the namespace version prior to pushing leaves onto the NMT so the NMT would be made aware of the version.

I agree option A2 makes sense. Added to doc in https://github.com/celestiaorg/celestia-app/pull/1382/commits/4a1489dde48a0c8a4c285ee13f679c27754fd022